vllm_hpu_extension/ops.py

```diff
@@ -347,34 +353,79 @@ def __init__(self, num_total_experts):
-                                                w12=w1_list,
-                                                w3=w2_list,
-                                                permuted_weights=permuted_weights,
-                                                activation=activation,
-                                                experts_min=0,
-                                                experts_max=self.num_experts - 1)
+        bt, hidden_dim = hidden_states.shape
+        num_experts = self.w13_weight.shape[0]
+        moe_intermediate = self.w2_weight.shape[2]
+        ep_shift = self.ep_rank * num_experts
+        selected_experts = (expert_routing_table - ep_shift).to(torch.int64)
+        if bt > dynamic_moe_min_tokens:
+            experts_range = range(num_experts)
+            w1_list = [self.w13_weight[i].squeeze() for i in experts_range]
+            w2_list = [self.w2_weight[i].squeeze() for i in experts_range]
+            final_hidden_states = torch.ops.hpu.mixture_of_experts(
+                hidden_states=hidden_states,
+                expert_routing_table=selected_experts,
+                router_weights=router_weights,
+                w12=w1_list,
+                w3=w2_list,
+                permuted_weights=True,
+                activation="silu",
+                experts_min=0,
+                experts_max=(num_experts - 1),
+            )
+        else:
+            padded_weights = torch.zeros((bt, num_experts),
+                                         dtype=hidden_states.dtype,
+                                         device=hidden_states.device)
+            padded_weights.scatter_(-1, selected_experts, router_weights)
+            padded_weights = padded_weights.transpose(0, 1).unsqueeze(-1)
 
+            up_gate_states = torch.matmul(
+                hidden_states,
+                self.w13_weight.view(-1, self.w13_weight.size(-1)).transpose(
+                    0, 1))
+            up_gate_states = up_gate_states.reshape(bt, num_experts, 2,
+                                                    moe_intermediate)
+            up_states = up_gate_states[:, :, 0, :]
+            gate_states = up_gate_states[:, :, 1, :]
+            current_state_static = F.silu(up_states) * gate_states
+            current_state_static = current_state_static.transpose(0, 1)
+
+            current_hidden_states_static = torch.matmul(
+                current_state_static, self.w2_weight.transpose(
+                    1, 2)) * padded_weights
+            final_hidden_states = current_hidden_states_static.sum(dim=0)
+
+        return final_hidden_states.view(-1, hidden_states.shape[1])
 
 class DynamicFusedMOE(torch.nn.Module):
 
     def __init__(self, num_total_experts):
         super().__init__()
         self.MoeOp = VllmMixtureOfExpertsOp(num_total_experts)
 
+    def set_MoeOp_weights(self, w13, w2):
+        self.MoeOp.set_weights(w13, w2)
+
+    def set_MoeOp_ep_rank(self, ep_rank):
+        self.MoeOp.set_ep_rank(ep_rank)
```

Comment:
Is that OK you can follow same approach as deepseek => since we have added necessary API changes in INC side and merged to 1.21 docker

for 1.21 (already merged) => https://github.com/HabanaAI/vllm-hpu-extension/pull/137
By using min_expert and max_expert

![image](https://github.com/user-attachments/assets/c0e73592-43ca-4e1c-a727-10ba40d4a4d0)